### DISCUSSION TITLE
CYFS automatically updates people information.

### DISCUSSION THREAD

**Discussion by wugren** [buckyos/CYFS#233] [Q&A]
My product needs to display the username and avatar of the people, but when the user updates the name and avatar, the people object obtained from CYFS is not updated. I believe many products have a similar need for this feature, and I hope CYFS can automatically update people information.

**Comment by lurenpluto** [comment]
In fact, this is an old problem, that is, the synchronization problem of objects in a weakly consistent system

The cyfs network based on device-zone is itself a "weak consistency network", that is, the consistency of updating and synchronization of objects inside the network is difficult to guarantee, so we have introduced the following object design rules.

# First, we clarify two concepts
## - Mutable objects
object with body, object-id will not change after body modification
## - Immutable objects
object without object, that is, the content stores in the desc content, if the content changes then object-id will change to a new object
	
# Two principles
## 1. Be careful to add and use mutable objects
The vast majority of objects should be immutable objects, so there is no consistency problem; to match this pattern, the introduction of root-state, you can request the corresponding data by `zone + req-path` (the path is fixed, but the corresponding data, that is, the object-id is subject to change), rather than based on the `object-id` to request ( This mode can only use mutable objects)
	
## 2. For mutable objects, use a strong consistency system to ensure
At present, it is guaranteed by the meta chain (block chain), for example, now the system inside the device and people object itself are mutable objects, and may change sometimes, so in order to ensure consistency, require the device and people object are stored in the meta chain, to ensure that the entire network the latest version of the object can be obtained with object-id.
	
If you don't use meta chain or similar centralized system to ensure the consistency of mutable objects, and only rely on the maintenance between zones, it will be very cumbersome and bring heavy burden
	
# So back to your requirement, I think we can consider it from the following two perspectives.

## 1. What information should be stored in the people object
If it is frequently changing information, or a large amount of data information, it will lead to high costs for people in the meta chain (publish and rent costs, etc.), so it should be more appropriate to store some information that changes less
	
For information that changes frequently and has a large amount of information, it is recommended to use `zone+req-path` to store it, so that you can query the corresponding information through` people+req-path` when permission allows; this requires a good definition of the path specification of people information in the root-state in advance, and we all follow this specification
	
## 2. Current cyfs-stack internal synchronous management of mutable objects (with body object)
At present, cyfs-stack actually lacks systematic synchronization of variable objects, so some synchronization mechanism should be designed to ensure the consistency of variable objects to some extent

**Comment by lurenpluto** [comment]
As for how to improve the consistency of mutable objects inside cyfs-stack, consider the following improvements:

- Track the source and last fetch time of mutable objects, such as meta chain, zone
- Periodically scan all mutable objects and try to initiate a synchronization operation for those beyond a certain time.
- Based on checkpointing mechanism, when a mutable object is read, check whether it needs to be updated.
- Design a suitable policy, such as user-configurable or depending on the type of mutable object to determine the refresh time
